src/main/java/org/jenkinsci/remoting/engine/WorkDirManager.java

```diff
@@ -0,0 +1,232 @@
+     * @param workDir Working directory
+     * @param internalDir Name of the remoting internal data directory within the working directory.
+     *                    The range of the supported symbols is restricted to {@link #SUPPORTED_INTERNAL_DIR_NAME_MASK}.
+     * @param failIfMissing Fail the initialization if the workDir or internalDir are missing.
+     *                      This option presumes that the workspace structure gets initialized previously in order to ensure that we do not start up with a borked instance
+     *                      (e.g. if a mount gets disconnected).
+     * @return Initialized directory for internal files within workDir or {@code null} if it is disabled
+     * @throws IOException Workspace allocation issue (e.g. the specified directory is not writable).
+     *                     In such case Remoting should not start up at all.
+     */
+    @CheckForNull
+    public Path initializeWorkDir(final @CheckForNull File workDir, final @Nonnull String internalDir, final boolean failIfMissing) throws IOException {
+
+        if (!internalDir.matches(SUPPORTED_INTERNAL_DIR_NAME_MASK)) {
+            throw new IOException(String.format("Name of %s ('%s') is not compliant with the required format: %s",
+                    DirType.INTERNAL_DIR, internalDir, SUPPORTED_INTERNAL_DIR_NAME_MASK));
+        }
+
+        if (workDir == null) {
+            // TODO: this message likely suppresses the connection setup on CI
+            // LOGGER.log(Level.WARNING, "Agent working directory is not specified. Some functionality introduced in Remoting 3 may be disabled");
+            return null;
+        } else {
+            // Verify working directory
+            verifyDirectory(workDir, DirType.WORK_DIR, failIfMissing);
+
+            // Create a subdirectory for remoting operations
+            final File internalDirFile = new File(workDir, internalDir);
+            verifyDirectory(internalDirFile, DirType.INTERNAL_DIR, failIfMissing);
+
+            // Create the directory on-demand
+            final Path internalDirPath = internalDirFile.toPath();
+            Files.createDirectories(internalDirPath);
+            LOGGER.log(Level.INFO, "Using {0} as a remoting working files directory", internalDirPath);
+            return internalDirPath;
+        }
+    }
+
+    /**
+     * Verifies that the directory is compliant with the specified requirements.
+     * The directory is expected to have {@code RWX} permissions if exists.
+     * @param dir Directory
+     * @param type Type of the working directory component to be verified
+     * @param failIfMissing Fail if the directory is missing
+     * @throws IOException Verification failure
+     */
+    private static void verifyDirectory(@Nonnull File dir, @Nonnull DirType type, boolean failIfMissing) throws IOException {
+        if (dir.exists()) {
```

Comment:
Hmm, I would expect the process to try to create the directory path if it is missing and complain if it can't create the directory.